### PR TITLE
APIv2: change signature of the 3-argument callback

### PR DESCRIPTION
**Pull request by aprokop**
Change the signature of the callback:
```diff
-callback(Predicate const &predicate, int primitive_index, OutputFunctor &functor);
+callback(Predicate const &predicate, Value const &value, OutputFunctor &functor);
```

### CODE REVIEW

File: src/ArborX_BruteForce.hpp [hunk 1]
```diff
@@ -123,21 +121,51 @@ class BruteForce
...
+
+      Kokkos::View<int *, MemorySpace> indices(
+          "ArborX::CrsGraphWrapper::query::indices", 0);
+      base_type::query(space, predicates, Details::DefaultLegacyCallback{},
+                       indices, std::forward<OffsetView>(offset),
+                       std::forward<Args>(args)...);
+      callback(predicates, std::forward<OffsetView>(offset), indices,
+               std::forward<OutputView>(out));
+    }
+  }
```
Comment (aprokop) [inline src/ArborX_BruteForce.hpp:168]: I moved the calls for the legacy classes out of `CrsGraphImpl`. The reason being is that the call stack was too complicated to figure out from within `CrsGraphImpl` when to wrap and when not to. Moving out made things doable.

The downside is that it had to be repeated in both `BruteForce` and `Testing::LegacyTree`. I think it's ok.

File: src/ArborX_LinearBVH.hpp [hunk 2]
```diff
@@ -172,23 +170,52 @@ class BoundingVolumeHierarchy
              Experimental::TraversalPolicy const &policy =
                  Experimental::TraversalPolicy()) const
   {
+    Details::check_valid_callback<int>(callback, predicates);
```
Comment (aprokop) [inline src/ArborX_LinearBVH.hpp:173]: Drive-by change

File: src/ArborX_BruteForce.hpp [hunk 3]
```diff
@@ -123,21 +121,51 @@ class BruteForce
   void query(ExecutionSpace const &space, Predicates const &predicates,
              Callback const &callback, Ignore = Ignore()) const
   {
+    Details::check_valid_callback<int>(callback, predicates);
```
Comment (aprokop) [inline src/ArborX_BruteForce.hpp:124]: Drive-by change

File: src/ArborX_BruteForce.hpp [hunk 4]
```diff
@@ -123,20 +121,52 @@ class BruteForce
...
   }
+
+  template <typename ExecutionSpace, typename Predicates, typename Callback,
+            typename OutputView, typename OffsetView, typename... Args>
+  std::enable_if_t<!Kokkos::is_view_v<std::decay_t<Callback>>>
+  query(ExecutionSpace const &space, Predicates const &predicates,
+        Callback &&callback, OutputView &&out, OffsetView &&offset,
+        Args &&...args) const
+  {
+    if constexpr (!Details::is_tagged_post_callback<
```
Comment (dalg24) [inline src/ArborX_BruteForce.hpp:146]: Just observing that we have this pattern 3x (BF, BVH, and LegacyTree)
Not sure whether we should/could abstract it away.

**Review by dalg24 — COMMENTED** [review]
Only minor comments

File: test/ArborXTest_LegacyTree.hpp [hunk 5]
```diff
@@ -27,6 +27,52 @@ class LegacyTree : public Tree
...
+    Tree::query(space, predicates, ArborX::Details::DefaultLegacyCallback{},
+                std::forward<View>(view), std::forward<Args>(args)...);
+  }
+
+  template <typename ExecutionSpace, typename Predicates, typename Callback,
+            typename OutputView, typename OffsetView, typename... Args>
+  std::enable_if_t<!Kokkos::is_view_v<std::decay_t<Callback>>>
+  query(ExecutionSpace const &space, Predicates const &predicates,
+        Callback &&callback, OutputView &&out, OffsetView &&offset,
+        Args &&...args) const
```
[the remark this answers is not in the document]
Comment (aprokop) [reply test/ArborXTest_LegacyTree.hpp:53]: Yes.

File: src/ArborX_LinearBVH.hpp [hunk 6]
```diff
@@ -172,22 +170,54 @@ class BoundingVolumeHierarchy
...
+      base_type::query(space, predicates,
+                       Details::LegacyCallbackWrapper<Callback>{
+                           std::forward<Callback>(callback)},
+                       std::forward<OutputView>(out),
+                       std::forward<OffsetView>(offset),
+                       std::forward<Args>(args)...);
+    }
+    else
+    {
+      KokkosExt::ScopedProfileRegion guard("ArborX::BVH::query_crs");
```
[the remark this answers is not in the document]
Comment (aprokop) [reply src/ArborX_LinearBVH.hpp:209]: No. This one does not go through the base class, so we have to add it here to retain the original stack.

File: src/ArborX_BruteForce.hpp (hunk 4 above)
Comment (aprokop) [reply src/ArborX_BruteForce.hpp:146]: Yes. Not ideal.

File: src/details/ArborX_DetailsLegacy.hpp [hunk 7]
```diff
@@ -66,6 +66,25 @@ struct LegacyCallbackWrapper
...
+  template <typename Predicate, typename Geometry, typename Output>
+  KOKKOS_FUNCTION auto operator()(Predicate const &predicate,
+                                  PairIndexVolume<Geometry> const &value,
+                                  Output const &out) const
+  {
+    return _callback(predicate, value.index, out);
+  }
+};
+
+struct DefaultLegacyCallback
```
[the remark this answers is not in the document]
Comment (aprokop) [reply src/details/ArborX_DetailsLegacy.hpp:79]: Good point. Now that you mention it, I like `LegacyDefaultCallback` better. Would change.

File: src/ArborX_LinearBVH.hpp [hunk 8]
```diff
@@ -172,22 +170,52 @@ class BoundingVolumeHierarchy
...
-                     std::forward<CallbackOrView>(callback_or_view),
+    base_type::query(space, predicates, Details::LegacyDefaultCallback{},
                      std::forward<View>(view), std::forward<Args>(args)...);
   }
 
+  template <typename ExecutionSpace, typename Predicates, typename Callback,
+            typename OutputView, typename OffsetView, typename... Args>
+  std::enable_if_t<!Kokkos::is_view_v<std::decay_t<Callback>>>
+  query(ExecutionSpace const &space, Predicates const &predicates,
+        Callback const &callback, OutputView &&out, OffsetView &&offset,
```
Comment (dalg24) [inline src/ArborX_LinearBVH.hpp:193]: If you take the callback by const reference then you don't need the `std::decay` below

File: src/ArborX_LinearBVH.hpp [hunk 9]
```diff
@@ -172,22 +170,52 @@ class BoundingVolumeHierarchy
...
+        Args &&...args) const
+  {
+    if constexpr (!Details::is_tagged_post_callback<
+                      std::decay_t<Callback>>::value)
+    {
+      Details::check_valid_callback<int>(callback, predicates, out);
+      base_type::query(
+          space, predicates, Details::LegacyCallbackWrapper<Callback>{callback},
```
Comment (dalg24) [inline src/ArborX_LinearBVH.hpp:201]: I think your real issue was here.  Were you getting memory access violations at runtime?  You'd need to decay the template parameter

File: src/ArborX_LinearBVH.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply src/ArborX_LinearBVH.hpp:207]: Yuk.  Not sure if that is worth a comment to keep them in sync.

**Review by dalg24 — COMMENTED** [review]
I think that it would be nicer to forward the callback

File: src/ArborX_LinearBVH.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (aprokop) [reply src/ArborX_LinearBVH.hpp:206]: I don't think that is important. We don't guarantee anything about profiling regions stacks. As long as we name them somewhat related, it should be fine.

File: src/ArborX_LinearBVH.hpp (hunk 9 above)
Comment (dalg24) [reply src/ArborX_LinearBVH.hpp:200]: Why did you go for dropping the std::decay instead of forwarding the callback and "fixing" the template argument?

Comment (aprokop) [reply src/ArborX_LinearBVH.hpp:200]: OK, I'll try. But can you explain what we gain by forwarding the callback?

Comment (dalg24) [reply src/ArborX_LinearBVH.hpp:200]: One less copy when passing a temporary (which doesn't really matter)
I mostly wanted to make sure we understand what was wrong with the initial version.
If you prefer const ref I will approve but would you please try locally so we know for sure.

Comment (aprokop) [reply src/ArborX_LinearBVH.hpp]: OK, you are right. I updated to decay the callback template argument and it worked.

**Review by dalg24 — APPROVED** [review]
